proposed/activity-improvements.md

```diff
@@ -0,0 +1,290 @@
+
+We expect to achieve this by additions to the Activity object model and a new ActivityListener type. These additions should provide stronger contracts between in-proc producers and consumers of distributed tracing data, a richer set of information that Activity can represent, and streamlined calling patterns for typical scenarios.
+
+## Scenarios and User Experience
+
+#### 1. A developer wants a portion of code to be annotated with a named block in a distributed trace Gantt chart
+
+The developer needs to wrap their region of code in a using statement that creates the Activity and assigns it a name. It is possible that the activity reference will be null if there is nothing in the process listening for this telemetry or if only a sub-set of telemetry is being sampled.
+
+```C#
+    using (Activity? activity = Activity.StartNew("MyLibrary.Subcomponent.FooOperation")
+    {
+        // optionally the activity can be enriched with extra data
+        activity?.AddTag("userId", userId);
+        activity?.AddTag("fooParam", foo);
+            
+        // the developer's original code goes here
+    }
+```
+
+### 2. Monitoring agent wants to listen to activities
+
+A monitoring agent can create a listener which represents their subscription to activity callbacks like this.
+```C#
+
+IDisposable listener = Activity.StartListening(
+           shouldCreateActivity: (name, parentContext, kind, tags, links) => true,
+           onActivityStarted: (activity) => { ... },
+           onActivityStopped: (activity) => { ... });
+
+```
+It is up to the listener to do extract information from the Activity object and respond to the callbacks however they like, probably by serializing a message to a remote service that will aggregate the telemetry.
+
+[TODO]: We could flesh out what we expect it look like end-to-end to use this with OpenTelemetry and perhaps some UI? It would help newcomers understand but I don't think it is a blocking issue for the current set of collaborators.
+
+Requirements:
+=============
+
+### Goals
+
+#### 1. All Activities can be collected
+
+The current design of DiagnosticListener and the conventions used by existing instrumented libraries expect the collection agent to have significant knowledge about each instrumented library. We need to make it possible for a general purpose collection agent to collect distributed trace telemetry despite having none of this advance knowledge. This also implies that any developer can add new Activity instrumentation to their app or library and the agent is capable or recording it automatically.
+
+#### 2. Activity collection can be filtered with a predicate on name
+
+We are assuming that most Activity information is desirable by default, but the monitoring agent should be able to filter out noisy or problematic instrumentation. Each Activity has a name and the agent should be able to efficiently decide if it wants the telemetry given the name.
```

Comment:
there is a problem with a single-tier names of activities. In OpenTelemetry there is a Tracer name and version and specific Activity name. Different versions of Tracers may produce an activity with the same name. But one of them may be desirable and another not. Also this design calls for the hierarchical naming convention for Activity while OpenTelemetry Span name is not. For instance, see the recommendation for the name of http spans: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name

Reply:
I think I have misunderstood how Tracer has evolved and perhaps got confused by more outdated info in the C# implementation. If I am following correctly [this text in the dotnet OT repo](https://github.com/open-telemetry/opentelemetry-dotnet#obtaining-tracer) is now outdated and I should ignore it?

> Libraries must take dependency on OpenTelemetry API package only and should never instantiate tracer or configure OpenTelemetry. Libraries will be able to obtain global tracer that may either be noop (if user application is not instrumented with OpenTelemetry) or real tracer implemented in the SDK package.

And the updated guidance [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#obtaining-a-tracer) suggests that each library should instantiate its own tracer like this: 

`Tracer t = tracerProvider.getTracer("some_string_identifying_my_library", version)`

Should I also assume that sampler implementations have access to the Tracer (or something Tracer derived) when they are constructed? The sampler receives the span name but if the names aren't unique then the information appears of little value if it can't be put in context? Maybe I missed it but I didn't find any information in the spec that describes how samplers are created/configured. I also didn't find any info about name based filtering that might occur distinct from samplers.

Assuming OT expects filtering to occur on a triplet of activity_name, component_name, and component_version I think we could find some remedies but as usual dotnet's historical baggage adds wrinkles:

1. Activity can always be constructed and started manually. This activity only has an OperationName and no other information. Either these Activities can be completely ignored or they need to be given sentinel values for component name/version.
2. Some activities can be started via DiagnosticSource.StartActivity() which logically provides a component name, however this name isn't known early enough to provide it to the sampling API and prevent the Activity from being created.

A potential solution:
1. Bring back a class called ActivitySource (it is somewhat like Tracer but there is enough difference I'd hesitate to name it that)
2. ActivitySource has properties called Name and Version. Unlike the earlier design iteration of ActivitySource this Name refers to the component and has nothing to do with the eventual activity name.
3. The APIs for Activity.StartNew(...) move to be ActivitySource.StartNew(...).
4. All Activities have a Source property indicating the source that started them. For activities that are created via Activity..ctor() they would get assigned to a global singleton source whose name is empty string and version is 0.0.0.0
5. ActivitySource shows up as one of the parameters in ShouldCreateActivity()
6. [Optionally] We do the perf optimization where listener lists are maintained per-source rather than globally. This requires another callback for the listener to filter which sources it wants to listen to. It also requires an internal weak-reference tracking list for sources if we don't want to force users to deal with Dispose()'ing them. If we are already paying the user mental model complexity of having ActivitySource for this naming purpose then I think it is a much lower bar to justify throwing in the perf optimization too.

This solution back-tracks on [@lmolkova's earlier request](https://github.com/dotnet/runtime/pull/32660#discussion_r382783881) to simplify and get rid of ActivitySource. However if we have to have a component name and version I think something like an ActivitySource object to encapsulate them is cleaner than asking developers to repeat the same name/version at every Activity creation site.

Reply:
cc @pakrym @benaadams on the perf part of this